R/edgeThresh.R

```diff
@@ -0,0 +1,284 @@
+
+    # After generating N vectors of size (Lmax - 1), find the median δ for each
+    # δ ord (maxComponentSize). i.e. Lmax - 1 median values.
+    deltaOrd <- vector("numeric", Lmax - 1)
+    for (i in 1:(Lmax - 1)) {
+        deltaOrd[i] <- median(sapply(masterDeltaList, FUN = "[", i))
+    }
+
+    # Attach the vector as graph metadata to the original EGG and return.
+    # (Is this the correct way?)
+    attributes(EGG) <- deltaOrd
+
+    # (Can this be parallelized? Since the processes are not interdependent)
+
+    # Write to Console:
+    # Update the fraction of N graphs that are done with permutations, (can
+    # also update the rough percentage of Q * |E| swaps that are done, but can
+    # slow down the process). This is done in .permute.
+    # Update the fraction of graphs that are done with finding δ vector (can
+    # also indicate the percentatge of maxComponentSize values that are
+    # finished per graph, but can again slow the process)
+
+    # Write to log:
+    # Log of passed arguments (function call record), processed information
+    # (number of edges and vertices, bidirectional and unidirectional)
+    # Output object (δ ord vector as metadata), start and finish times
+    # (per graph, or for the whole process?)
+}
+
+
+
+#' Helper function to create N permutations of the EGG graph.
+#'
+#' @param EGG The given EGG graph from parent function.
+#' @param N The number of permuted graphs that must be generated.
+#' @param Q The number of edge-swapping permutations to produce one of the
+#' permuted networks is Q * |E|.
+#'
+#' @return List of N igraph objects, permutations of the given EGG graph
+.permute <- function(EGG, N, Q) {
+    # Calculate the component each vertex belongs to
+    ### By subsetting components (somehow)
+    initialComponents <- igraph::components(EGG)
+    numComponents = initialComponents$no
+
+
+    # Create a hash table to track components vertecies
+    # to. (can a package be used? fastmatch? or just hashed env?)
+    # Assign cluster ID to each node
+    # Vertex names must be unique
+    componentTracker <- new.env(hash = TRUE, size = numVertices)
+    for (i in 1:numVertices) {
+        # Should $name be used?
+       componentTracker[[allVertices$name[i]]] <- initialComponents$membership[i]
+    }
+
+    # Get a list of all edges, separate find bidirectional ones
+    # Courtesy of Sacha Epskamp,
+    E <- t(apply(X = allEdges, MARGIN = 1, FUN = sort))
+    bidirectionalEdges <- allEdges[duplicated(E) | duplicated(E, fromLast = TRUE)]
+    numBidirectional <- (length(bidirectionalEdges)) / 2
+    # Since they are all ordered, we can pick only the first half, and group
+    # each pair
+
+    # Better to use a list of tuples to keep track of 'bidirectional' edge pairs
+    # as they are simply a pair of unidirectional edges
+    bidirPairList <- vector("list", length = (numBidirectional / 2))
+    for (i in seq(1, numBidirectional, 2)) {
```

Comment:
You don't need that. Bidirectionality is implied. Separate just one of each pair out into bidirected category, and whenever a swap is required, swap both.